plugin/validator/validator.go

```diff
@@ -0,0 +1,295 @@
+In case some of the fields are annotated, the Validate function returns nil upon sucessful validation, or an error
+describing why the validation failed.
+The Validate method is called recursively for all submessage of the message.
+
+TODO(michal): ADD COMMENTS.
+
+Equal is enabled using the following extensions:
+
+  - equal
+  - equal_all
+
+While VerboseEqual is enable dusing the following extensions:
+
+  - verbose_equal
+  - verbose_equal_all
+
+The equal plugin also generates a test given it is enabled using one of the following extensions:
+
+  - testgen
+  - testgen_all
+
+Let us look at:
+
+  github.com/gogo/protobuf/test/example/example.proto
+
+Btw all the output can be seen at:
+
+  github.com/gogo/protobuf/test/example/*
+
+The following message:
+
+
+
+given to the equal plugin, will generate the following code:
+
+
+
+and the following test code:
+
+
+*/
+package validator
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/gogo/protobuf/gogoproto"
+	descriptor "github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"github.com/gogo/protobuf/protoc-gen-gogo/generator"
+)
+
+type plugin struct {
+	*generator.Generator
+	generator.PluginImports
+	regexPkg generator.Single
+	fmtPkg   generator.Single
+	protoPkg generator.Single
+}
+
+func NewPlugin() *plugin {
+	return &plugin{}
+}
+
+func (p *plugin) Name() string {
+	return "validator"
+}
+
+func (p *plugin) Init(g *generator.Generator) {
+	p.Generator = g
+}
+
+func (p *plugin) Generate(file *generator.FileDescriptor) {
+	p.PluginImports = generator.NewPluginImports(p.Generator)
+	p.regexPkg = p.NewImport("regexp")
+	p.fmtPkg = p.NewImport("fmt")
+
+	for _, msg := range file.Messages() {
+		if msg.DescriptorProto.GetOptions().GetMapEntry() {
+			continue
+		}
+		p.generateRegexVars(file, msg)
+		if gogoproto.IsProto3(file.FileDescriptorProto) {
+			p.generateProto3Message(file, msg)
+		} else {
+			p.generateProto2Message(file, msg)
+		}
+
+	}
+}
+
+func (p *plugin) isSupportedInt(field *descriptor.FieldDescriptorProto) bool {
+	switch *(field.Type) {
+	case descriptor.FieldDescriptorProto_TYPE_INT32, descriptor.FieldDescriptorProto_TYPE_INT64:
+		return true
+	case descriptor.FieldDescriptorProto_TYPE_UINT32, descriptor.FieldDescriptorProto_TYPE_UINT64:
+		return true
+	case descriptor.FieldDescriptorProto_TYPE_SINT32, descriptor.FieldDescriptorProto_TYPE_SINT64:
+		return true
+	}
+	return false
+}
+
+func (p *plugin) generateRegexVars(file *generator.FileDescriptor, message *generator.Descriptor) {
+	ccTypeName := generator.CamelCaseSlice(message.TypeName())
+	for _, field := range message.Field {
+		validator := gogoproto.GetValidator(field)
+		if validator != nil && validator.Regex != nil {
+			fieldName := p.GetFieldName(message, field)
+			p.P(`var `, p.regexName(ccTypeName, fieldName), ` = `, p.regexPkg.Use(), `.MustCompile("`, validator.Regex, `")`)
+		}
+	}
+}
+
+func (p *plugin) generateProto2Message(file *generator.FileDescriptor, message *generator.Descriptor) {
```

Comment:
You should be able to fold proto2 and proto3 into one method generateMethod with some ifs for proto3, check some of the other plugins.